core/src/test/java/org/infinispan/globalstate/AbstractGlobalStateRestartTest.java

```diff
@@ -0,0 +1,212 @@
+
+   public static int DATA_SIZE = 1000;
+
+   @Override
+   protected boolean cleanupAfterMethod() {
+      return true;
+   }
+
+   @Override
+   protected boolean cleanupAfterTest() {
+      return false;
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      TestingUtil.recursiveFileRemove(TestingUtil.tmpDirectory(this.getClass().getSimpleName()));
+      createStatefulCacheManagers(true, -1, false);
+   }
+
+   protected void createStatefulCacheManagers(boolean clear, int extraneousNodePosition, boolean reverse) throws Throwable {
+      int totalNodes = getClusterSize() + ((extraneousNodePosition < 0) ? 0 : 1);
+      int node = reverse ? getClusterSize() - 1 : 0;
+      int step = reverse ? -1 : 1;
+      for (int i = 0; i < totalNodes; i++) {
+         if (i == extraneousNodePosition) {
+            // Create one more node if needed in the requested position
+            createStatefulCacheManager(Character.toString('@'), true);
+         } else {
+            createStatefulCacheManager(Character.toString((char) ('A' + node)), clear);
+            node += step;
+         }
+      }
+   }
+
+   private void createStatefulCacheManager(String id, boolean clear) {
+      String stateDirectory = TestingUtil.tmpDirectory(this.getClass().getSimpleName() + File.separator + id);
+      if (clear)
+         TestingUtil.recursiveFileRemove(stateDirectory);
+      GlobalConfigurationBuilder global = GlobalConfigurationBuilder.defaultClusteredBuilder();
+      global.globalState().enable().persistentLocation(stateDirectory);
+
+      ConfigurationBuilder config = new ConfigurationBuilder();
+      applyCacheManagerClusteringConfiguration(config);
+      config.persistence().addSingleFileStore().location(stateDirectory);
+      addClusterEnabledCacheManager(global, config);
+   }
+
+   protected abstract void applyCacheManagerClusteringConfiguration(ConfigurationBuilder config);
+
+   protected void shutdownAndRestart(int extraneousNodePosition, boolean reverse) throws Throwable {
+      Map<JGroupsAddress, PersistentUUID> addressMappings = createInitialCluster();
+
+      ConsistentHash oldConsistentHash = cache(0).getAdvancedCache().getDistributionManager().getConsistentHash();
+
+      // Shutdown the cache cluster-wide
+      cache(0).shutdown();
+
+      TestingUtil.killCacheManagers(this.cacheManagers);
+
+      // We should have some data here
+      for (int i = 0; i < getClusterSize(); i++) {
+         checkStateDirNotEmpty(manager(i).getCacheManagerConfiguration().globalState().persistentLocation());
+      }
+      this.cacheManagers.clear();
+
+      // Simulate a restart
+      JGroupsAddressCache.flushAddressCaches();
+
+      // Recreate the cluster
+      createStatefulCacheManagers(false, extraneousNodePosition, reverse);
+      if(reverse) {
+         Map<JGroupsAddress, PersistentUUID> reversed = new LinkedHashMap<>();
+         reverseLinkedMap(addressMappings.entrySet().iterator(), reversed);
```

Comment:
Couldn't we rely on the node name instead? Depending on the ordering seems a bit fragile to me.